### PR TITLE
Complete MoA boundary validation and VisDrone convergence report

### LINKED ISSUE #53: 2026犀牛鸟开源人才专属】【低难度】MoA(mixture of Attention)：边界测试补全与垂类训练验证
在现有 tests/test_moa.py 基础上，系统补全 MoA 模块的边界测试与训练验证，提升测试覆盖率
测试补全（至少覆盖以下场景）：
NeckMoAFusion 在跨尺度输入尺寸不匹配（如 hi 为 15×15、lo 为 7×7，非严格 2× 下采样）时的前向稳定性与形状保持
MoABlock 的 temperature 退火到极小值（如 temperature < 1e-4）时，softmax 路由概率的数值稳定性（是否出现 NaN 或均匀分布）
_LocalAttnHead 与 _GlobalAttnHead 在 num_heads 不能被 dim 整除时的降级处理（_safe_groups 的边界）
C2fMoA 的 aux_loss 在多 MoABlock 嵌套时是否存在重复计数（类似 MoE 的 MOE_LOSS_REGISTRY 双计数问题）
缺陷修复：在补充测试过程中，若发现任何边界缺陷（如 IndexError、NaN 传播、形状不匹配），需一并定位并修复
覆盖率报告：提供 pytest --cov 前后的覆盖率对比，至少覆盖 ultralytics/nn/modules/moa/ 目录
垂类训练验证：在 VisDrone 或 SKU-110K 上，使用 YOLO-Master-v0.10-MoA-N 训练 50~100 epoch，验证 MoA 模块在真实数据集上的收敛性，记录 mAP50-95 与 loss 曲线，与同配置的 MoE 基线对比
提交 Pull Request，包含测试代码、修复代码、覆盖率报告、训练日志
https://github.com/Tencent/YOLO-Master
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moa/moa.py
https://github.com/Tencent/YOLO-Master/blob/main/tests/test_moa.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/cfg/models/master/v0_10/det/yolo-master-moa-n.yaml
https://github.com/Tencent/YOLO-Master/blob/main/scripts/compare_moa_ablation.py
难度：低

---

**本issue为2026犀牛鸟开源人才培养活动专属issue，仅供已报名参与犀牛鸟活动的同学认领**
【认领时间】7月1日～7月31日（7月1日前认领视为无效❗️）
【认领方式】在本issue评论区回复“已认领本任务”，即视为认领成功
【活动报名】需提前完成犀牛鸟报名问卷，问卷将用于活动登记和奖励发放：https://wj.qq.com/s2/26888567/gh2q

### PR DESCRIPTION
**Pull request by kimariyb**
Closes #53 

## Summary

This PR completes the MoA boundary validation and vertical training verification work for `test/moa-boundary-vertical-validation`.

It adds systematic boundary tests for `ultralytics/nn/modules/moa/`, fixes validation-time helper issues, and documents the coverage/training results under `examples/YOLOv10-Master-MoA/`.

  ## Changes

  ### MoA Boundary Tests

  Added and expanded tests in `tests/test_moa.py` for:

  - `NeckMoAFusion` with non-strict cross-scale inputs, e.g. `hi=15x15`, `lo=7x7`
  - `MoABlock` router stability when temperature is extremely small, e.g. `< 1e-4`
  - `_LocalAttnHead` and `_GlobalAttnHead` fallback behavior when `num_heads` does not divide `dim`
  - `C2fMoA` aux loss collection to prevent duplicated nested `MoABlock` losses
  - `_flash_attn` fallback path when SDPA is unavailable
  - `_GlobalAttnHead` large-map linear attention path
  - `_MoARouter` probability-only return path
  - `MoABlock(shortcut=False)` branch
  - `NeckMoAFusion` eval-mode projection and zero-aux behavior
  - Empty-module and standalone `MoABlock` aux-loss collection

  ### Fixes

  - Fixed `NeckMoAFusion` cross-scale feature alignment by resizing the low-resolution feature map to `hi.shape[2:]`, preventing shape mismatch on odd or non-2x feature sizes.
  - Fixed `BaseValidator` final validation path by importing `LOCAL_RANK` and `torch_distributed_zero_first`.
  - Added `convert_ndjson_to_yolo_if_needed()` so standalone validation keeps YAML inputs unchanged and converts NDJSON inputs when needed.

  ### Documentation and Results

  Added `examples/YOLOv10-Master-MoA/` with:

  - `README.md`
  - coverage report
  - VisDrone 50-epoch vertical training comparison
  - model build summary
  - scalar result CSVs
  - seaborn mAP50-95 and loss-curve plots
  - reproducible plotting script

  ## Coverage

  Coverage was measured with:

  ```bash
  python -m pytest tests/test_moa.py tests/test_mixture_aux_loss.py \
    --cov=ultralytics/nn/modules/moa \
    --cov-report=term-missing \
    --cov-report=html:runs/moa_ablation/coverage_html -q
```

Before/after comparison:

| Stage | Test files | Statements | Missed | Coverage | Passed tests |
| --- | --- | ---: | ---: | ---: | ---: |
| Main branch baseline (`9163e96`) | `tests/test_moa.py`, `tests/test_mixture_aux_loss.py` | 297 | 17 | 94% | 9 |
| Current branch after boundary completion | `tests/test_moa.py`, `tests/test_mixture_aux_loss.py` | 297 | 0 | 100% | 20 |

Final terminal coverage summary:

```text
Name                                     Stmts   Miss  Cover
----------------------------------------------------------------------
ultralytics/nn/modules/moa/__init__.py       2      0   100%
ultralytics/nn/modules/moa/moa.py          295      0   100%
----------------------------------------------------------------------
TOTAL                                      297      0   100%
```

  ## Vertical Training Validation

  Training setup:

| Setting | Value |
| --- | --- |
| Dataset | `ultralytics/cfg/datasets/VisDrone.yaml` |
| Epochs | 50 |
| Image size | 640 |
| Batch size | 24 |
| Device | `0` |
| AMP | `False` |
| Deterministic | `True` |
| Project | `runs/moa_ablation/visdrone_v10_50ep` |

  Compared models:

| Role         | Model key | Config                                                         |
|--------------|-----------|----------------------------------------------------------------|
| MoE baseline | v10       | ultralytics/cfg/models/master/v0_10/det/yolo-master-n.yaml     |
| MoA variant  | v10_moa   | ultralytics/cfg/models/master/v0_10/det/yolo-master-moa-n.yaml |

  Final results:

| Metric                  | YOLO-Master v0.10 baseline       | YOLO-Master v0.10 MoA            |
|-------------------------|----------------------------------|----------------------------------|
| Params                  | 3,449,963                        | 3,576,587                        |
| mAP50                   | 0.22113                          | 0.22018                          |
| mAP50-95                | 0.12119                          | 0.12104                          |
| Precision               | 0.31388                          | 0.32180                          |
| Recall                  | 0.24208                          | 0.23843                          |
| Train Loss box/cls/dfl  | 1.54358 / 1.19714 / 0.97160      | 1.55375 / 1.20763 / 0.97522      |
| Val Loss box/cls/dfl    | 1.54373 / 1.19511 / 0.95679      | 1.54962 / 1.19342 / 0.95856      |
| Training Time           | 2.22 h                           | 2.59 h                           |

  Observation:

  - MoA converged successfully on VisDrone for 50 epochs.
  - Final mAP50-95 is nearly tied with the MoE baseline: -0.00015 absolute.
  - MoA precision is slightly higher.
  - MoA recall is slightly lower.
  - MoA adds 126,624 parameters and takes about 22.26 more minutes in this run.

### CODE REVIEW

**Review by isLinXu — COMMENTED** [review]
LGTM
